app/controllers/api/storage.php

```diff
@@ -927,41 +1044,95 @@
+                }
+                $rangeStart = (int) $rangeStart;
+                if(strlen($rangeEnd) == 0) {
+                    $rangeEnd =  min(($rangeStart + 2000000-1), ($size - 1));
+                } else {
+                    $rangeEnd = (int) $rangeEnd;
+                }
+                if(($rangeStart >= $rangeEnd) || $rangeEnd >= $size) {
+                    throw new Exception('Invalid range', 416);
+                }
+                
+                $response
+                    ->addHeader('Accept-Ranges', 'bytes')
+                    ->addHeader('Content-Range', 'bytes ' . $rangeStart . '-' . $rangeEnd . '/' . $size)
+                    ->addHeader('Content-Length', $rangeEnd - $rangeStart + 1)
+                    ->setStatusCode(Response::STATUS_CODE_PARTIALCONTENT);
+            } else {
+                throw new Exception('Invalid range', 416);
+            }
+        }
+
+        $source = '';
         if (!empty($file->getAttribute('openSSLCipher'))) { // Decrypt
+            $source = $device->read($path);
             $source = OpenSSL::decrypt(
                 $source,
                 $file->getAttribute('openSSLCipher'),
-                App::getEnv('_APP_OPENSSL_KEY_V'.$file->getAttribute('openSSLVersion')),
+                App::getEnv('_APP_OPENSSL_KEY_V' . $file->getAttribute('openSSLVersion')),
                 0,
                 \hex2bin($file->getAttribute('openSSLIV')),
                 \hex2bin($file->getAttribute('openSSLTag'))
             );
         }
-        if(!empty($file->getAttribute('algorithm', ''))) {
+
+        if (!empty($file->getAttribute('algorithm', ''))) {
+            if(empty($source)) {
+                $source = $device->read($path);
+            }
             $compressor = new GZIP();
             $source = $compressor->decompress($source);
         }
 
-        $usage
-            ->setParam('storage.files.read', 1)
-            ->setParam('bucketId', $bucketId)
-        ;
+        if(!empty($source)) {
+            if(!empty($rangeHeader)) {
+                $response->send(substr($source, $rangeStart, ($rangeEnd - $rangeStart + 1)));
+            }
+            $response->send($source);
+        }
 
-        // Response
-        $response
-            ->setContentType($file->getAttribute('mimeType'))
-            ->addHeader('Content-Disposition', 'attachment; filename="'.$file->getAttribute('name', '').'"')
-            ->addHeader('Expires', \date('D, d M Y H:i:s', \time() + (60 * 60 * 24 * 45)).' GMT') // 45 days cache
-            ->addHeader('X-Peak', \memory_get_peak_usage())
-            ->send($source)
-        ;
+        if(!empty($rangeHeader)) {
+            $response->send($device->read($path, $rangeStart, ($rangeEnd - $rangeStart + 1)));
+        }
+
+        if ($size > APP_STORAGE_READ_BUFFER) {          
+            $response->addHeader('Content-Length', $device->getFileSize($path));
+            $chunk = 2000000; // Max chunk of 2 mb
```

Comment:
Can we use a constant here? 🙂

Reply:
What would i name it, it's required in 2 places, 
1, when range header doesn't pass the rangeEnd and when we are just sending chunk of file as stream

Reply:
Something like `MAX_OUTPUT_CHUNK_SIZE`?